### PR TITLE
Estimate  days left for next purchase  when marking an item as purchased

### LINKED ISSUE #10: 10. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items.
AC:

* When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

The script at `/src/lib/estimates.js` should be used to calculate the next estimated purchase interval.

### PR DESCRIPTION
**Pull request by nibble0101**
## Description

This code computes the days  till next purchase when a user is marking an item as purchased. We have refactored the  shape of entries in `items` array in the `firestore` cloud storage to have the following fields:

```js
  {
      shoppingListItemName: string,
      daysLeftForNextPurchase: number[],
      purchaseDates: number[],
      numberOfPurchases: number,
   }
```
This decision was taken because it is important to keep track of `daysLeftForNextPurchase` and  ` purchaseDates` because of the parameters needed to compute `daysLeftForNextPurchase` using `calculateEstimate` function and the need to revert if a user marks an item purchased by mistake. Essentially `calculateEstimate` has the following parameters:
```js
 @param {Number} lastEstimate The last estimated purchase interval
 @param {Number} latestInterval The number of days between the most recent and previous purchases
 @param {Number} numberOfPurchases Total number of purchases for the item
```
### First purchase
For the first purchase, the value of `lastEstimate` is `undefined` because there is no last estimated purchase interval. The value of `latestInterval ` is either `7`, `14` or `30`  if the use thinks he/she will purchase the item "Soon", "Kind of soon", "Not soon" at the time of adding the item to the shopping list and the ` numberOfPurchases` is `1`. Passing these arguments to ``calculateEstimate`` returns a value which is  stored in `daysLeftForNextPurchase` array. We also store the current date, obtained from `Date.now()`,  in `purchaseDates` array.

### Second purchase
For the second purchase, the value of `lastEstimate` is the value of  `daysLeftForNextPurchase` estimated in the previous purchase. The value of `latestInterval ` is got by calculating the number days from the date the last purchase was made till now, when one is recording second purchase. The value of ` numberOfPurchases` now becomes `2`. These values are passed to `calculateEstimate` to estimate days left for next purchase. If a user realizes he/she has made a mistake in marking an item as purchased and decides to rescind that decision, they can do so because we are tracking purchase dates and days left for next purchase for previous and current purchase. If we don't persist  the values for the current purchase till the next purchase, it wouldn't be possible to revert if a user marks an item purchased by mistake.

Since we are only interested in persisting the values of  `daysLeftForNextPurchase`and `purchaseDate` for previous and current purchases, we have optimized the code to store a maximum of two entries in each of those two arrays.

## Related Issue

Closes #10 

## Acceptance Criteria

- When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database. The script at `/src/lib/estimates.js` should be used to calculate the next estimated purchase interval.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

image

### After

image

### Warning
The code in this feature will not work with token generated from previous versions of the app because of changes in the document structure in firestore cloud storage. First delete the  token already in your local storage and begin with a clean slate.

1. Switch to the feature branch using `git checkout -b cu-jm-estimate-next-purchase-date`
2. Pull the feature branch to your local setup using `git pull origin cu-jm-estimate-next-purchase-date`
3. Run `npm run start` which will launch the app on `localhost`, port 3000 if there is nothing else running on that port.
4. You can create a shopping list and add an  item to the shopping list
5. Mark the item as purchased
6. You can change the first element in the `purchaseDates` array to a date 5 days ago i.e `Date.now() - 5 * 24 * 60 * 60 * 1000`. 
7. Mark the item purchased again. Now change the second element in the `purchaseDates` array to a date  4 days ago i.e. 
`Date.now() - 4 * 24 * 60 * 60 * 1000`. Notice how the entries in `daysLeftForNextPurchase` array changes. You can continue playing with the code. 

***Note***
Adding  random previous  purchase dates will give you strange values of  `daysLeftForNextPurchase`

### CODE REVIEW

File: src/components/ItemsList.js
Comment (YuriDevAT) [inline src/components/ItemsList.js:56]: This explanation is amazing!!

**Review by YuriDevAT — COMMENTED** [review]
The explanations for how you did write your code are amazing. To give the user the possibility to uncheck an item when checked by accident and still get the right purchased date is really well written and a good feature! The code is long, but because of the good naming of the `const`, it is still easy to read.

File: src/components/ItemsList.js
Comment (YuriDevAT) [inline src/components/ItemsList.js:19]: This is pretty cool, that you changed it to a string to store all the puchaseDates! 😮

Comment (Amaka202) [reply src/components/ItemsList.js:19]: yea, its rather an array keeping track of last 2 purchase dates

File: src/components/ItemsList.js
Comment (rb50) [inline src/components/ItemsList.js:43]: I have a feeling that 'undefined' shouldn't be passed to a function/assigned to a variable etc. as it generally means that a variable/argument has been declared but not passed any value, whereas "null" means there is no value, so I think null would be more appropriate here?

File: src/components/ItemsList.js
Comment (rb50) [inline src/components/ItemsList.js:48]: Very tiny suggestion! But I think it's possible for line 47 to state } else if (....) { just removing the need for the if statement being on a separate line.

File: src/components/ItemsList.js
Comment (rb50) [inline src/components/ItemsList.js:53]: Again a really small suggestion but since we use the Date.now(); also on line 40, maybe it would be more readable assign the 'now' variable at the top of the function (line 38) and then it can be reused in line 40?

File: src/components/ItemsList.js
Comment (rb50) [inline src/components/ItemsList.js:37]: The logic within this function is really great, it's so easy to follow and review!

File: src/components/ItemsList.js
Comment (salmanmanekia) [reply src/components/ItemsList.js:43]: I think that the lastPurchasedOn can never be null as when the user adds an item then the purchased date can be added and when the user marks it for re-purchase then the lastPurchasedOn date can be used.

File: src/components/ItemsList.js
Comment (salmanmanekia) [inline src/components/ItemsList.js:39]: I think the lastPurchasedOn date can never be null. It can be set once the user adds an item to the list and then on the subsequent purchase that value can be used. I could be wrong but thats how I see it. 😄

File: src/components/ItemsList.js
Comment (alexlsalt) [reply src/components/ItemsList.js:19]: This makes sense! Also, good variable naming 👍

File: src/components/ItemsList.js
Comment (alexlsalt) [reply src/components/ItemsList.js:43]: I would agree here with declaring null instead of undefined 👍

**Review by alexlsalt — COMMENTED** [review]
Great work, Joseph and Chiamaka! I'm going to wait on the other mentors for final approval, but this looks good. Lots of thoughtful code and documentation - love to see that. 

Looking forward to our Sunday sync where we can see your demos! Proud of your work 🤗

File: src/components/ItemsList.js
Comment (Amaka202) [reply src/components/ItemsList.js:39]: ok thanks for the review, the lastPurchasedOn is indeed null until a user makes first purchase(when they check an item). They can add an item as a todo list without outrightly buying the item, lastPurchasedOn is null at that point, then when they tick the item as bought, lastPurchasedOn will be recorded as the time when they ticked the box.

Comment (Lukabaramishvili) [reply src/components/ItemsList.js:39]: Yes, I agree @Amaka202. Initially, the `lastPurchasedOn` should be null, as it also was part of the previous issue's AC.

File: src/components/ItemsList.js
Comment (Lukabaramishvili) [reply src/components/ItemsList.js:43]: Can you please be kind and explain why are we using `undefined` here? I'm not sure if I'm following the logic of why we are using `calculateEstimate` function in the first `if` statement block. 🙏 

I totally get the second else block from line 50 - 59. Thank you!

File: src/components/ItemsList.js
Comment (Lukabaramishvili) [inline src/components/ItemsList.js:59]: This is awesome. `calculateEstimate` function has all three @params: 
- last estimate purchase interval (in our case named as: `daysLeftForNextPurchase`)
- The number of days between the most recent and previous purchases (in our case named as: `getDaysBetweenCurrentAndPreviousPurchase `)
- Total number of purchases for the item (in our case named as: `numberOfPurchases`)

Great implementation! 🙌

**Review by Lukabaramishvili — APPROVED** [review]
Amazing job team! Everything works as expected and your PR meets all the AC's. I just left one question and comment. 

Absolutely amazing by the two versions of implementations. Both versions worked like a charm. 
Thank you again! We are all proud of you! 👏 🚀

File: src/components/ItemsList.js
Comment (nibble0101) [reply src/components/ItemsList.js:59]: Thanks @Lukabaramishvili for the review. Essentially for the first purchase, calling `calculateEstimate` is not necessary at all because the function  returns the same value passed if `numberOfPurchases < 2`. I am surprised you caught it. We shall remove it in our final follow up to the review comments.

Comment (Lukabaramishvili) [reply src/components/ItemsList.js:59]: @nibble0101 Sounds great. Thank you, Joseph.

File: src/components/ItemsList.js
Comment (nibble0101) [reply src/components/ItemsList.js:43]: Thanks @rb50  and @alexlsalt  for pointing that out. Actually invoking that function is not necessary at that point because it returns the same value passed. We shall have to remove it entirely.

File: src/components/ItemsList.js
Comment (nibble0101) [reply src/components/ItemsList.js:53]: Thanks for the suggestion. We can do that for readability though we wanted the recorded time to be as close to the time the data is written to the database as possible. Essentially given the trajectory the app is taking, we shall have to add more lines of code in that function. If we get `Date.now` and take sometime executing many lines of code, our recorded time may not be as precise.

File: src/components/ItemsList.js
Comment (YuriDevAT) [reply src/components/ItemsList.js:19]: Great changes here in variable naming.

File: src/components/ItemsList.js
Comment (Amaka202) [reply src/components/ItemsList.js]: Thank you for the reviews, we have taken out the function call entirely

File: src/components/ItemsList.js
Comment (Amaka202) [reply src/components/ItemsList.js:45]: thanks Becky, we structured it in such a way that the if else will run only when the condition for else is true. and also there's another if statement that will run if the other if evaluates to true. so for readability we separated them that way

File: src/components/ItemsList.js
Comment (Amaka202) [reply src/components/ItemsList.js]: We have now created a variable that calls the Date.now() and holds the value for reusability

File: src/components/ItemsList.js
Comment (salmanmanekia) [reply src/components/ItemsList.js:41]: That's right. Thanks for the clarification 👍🏻

File: src/components/ItemsList.js
Comment (faysvas) [inline src/components/ItemsList.js:20]: Does currentPurchaseDate need to be passed as a parameter? Maybe you can calculate it inside the `getDaysBetweenCurrentAndPreviousPurchase` function :thinking: ?

**Review by faysvas — APPROVED** [review]
Great work and fantastic explanation of your code review :100:! Functionality wise this looks good. I have a minor code suggestion that you can ignore if you're pressed for time. As I wrote in the other PR, I was a bit late to code review :sweat_smile:

**Review by salmanmanekia — APPROVED** [review]

**Pull request merged**
